ArcGISRuntimeSDKQt_CppSamples/Analysis/ConfigureSubnetworkTrace/ConfigureSubnetworkTrace.cpp

```diff
@@ -0,0 +1,368 @@
+  emit expressionBuilderChanged();
+}
+
+void ConfigureSubnetworkTrace::trace()
+{
+  m_busy = true;
+  emit busyChanged();
+  if (!m_utilityNetwork || !m_utilityElementStartingLocation)
+  {
+    return;
+  }
+  else
+  {
+    const QList<UtilityElement*> startingLocations {m_utilityElementStartingLocation};
+    // Create utility trace parameters for the starting location.
+    m_traceParams = new UtilityTraceParameters(UtilityTraceType::Subnetwork, startingLocations, this);
+    m_traceParams->setTraceConfiguration(m_traceConfiguration);
+
+    // trace the network
+    m_utilityNetwork->trace(m_traceParams);
+  }
+}
+
+void ConfigureSubnetworkTrace::onTraceCompleted()
+{
+  if (m_utilityNetwork->traceResult()->isEmpty())
+  {
+    m_dialogText = QString("No results returned");
+    emit dialogTextChanged();
+    m_dialogVisible = true;
+    emit dialogVisibleChanged();
+  }
+  // Get the first result.
+  UtilityTraceResult* result = m_utilityNetwork->traceResult()->at(0);
+
+  const QList<UtilityElement*> elements = static_cast<UtilityElementTraceResult*>(result)->elements(this);
+
+  // Display the number of elements found by the trace.
+  m_dialogText = QString("%1 elements found.").arg(elements.length());
+  m_dialogVisible = true;
+  m_busy = false;
+  emit dialogTextChanged();
+  emit dialogVisibleChanged();
+  emit busyChanged();
+}
+
+void ConfigureSubnetworkTrace::onUtilityNetworkLoaded(const Error& e)
+{
+  if (!e.isEmpty())
+  {
+    m_dialogText = QString("%1 - %2").arg(e.message(), e.additionalMessage());
+    m_dialogVisible = true;
+    emit dialogTextChanged();
+    emit dialogVisibleChanged();
+    return;
+  }
+
+  m_busy = false;
+  emit busyChanged();
+  m_networkDefinition = m_utilityNetwork->definition();
+
+  // Build the choice lists for network attribute comparison.
+  for (UtilityNetworkAttribute* networkAttribute : m_networkDefinition->networkAttributes())
+  {
+    if (!networkAttribute->isSystemDefined())
+      m_attributeListModel.append(networkAttribute->name());
+  }
+  emit attributeListModelChanged();
+
+  // Create a default starting location.
+  const UtilityNetworkSource* networkSource = m_networkDefinition->networkSource(m_deviceTableName);
+  const UtilityAssetGroup* assetGroup = networkSource->assetGroup(m_assetGroupName);
+  UtilityAssetType* assetType = assetGroup->assetType(m_assetTypeName);
+  m_utilityElementStartingLocation = m_utilityNetwork->createElementWithAssetType(assetType, m_gloabId);
+
+  QList<UtilityTerminal*> terminals = m_utilityElementStartingLocation->assetType()->terminalConfiguration()->terminals();
+
+  // Set the terminal for this location. (For our case, we use the 'Load' terminal.)
+  for (UtilityTerminal* terminal : terminals)
+  {
+    if (terminal->name() == "Load")
+    {
+      m_utilityElementStartingLocation->setTerminal(terminal);
+      break;
+    }
+  }
```

Comment:
Optional, consider [std::find_if](https://en.cppreference.com/w/cpp/algorithm/find) here. If that results in even more code or more complicated code - keep what you have.